site/source/_posts/2017/11/17-Dojo-2-Beta-4/Dojo-2-Beta-4.md

```diff
@@ -0,0 +1,263 @@
+
+## Integrated virtual DOM
+
+This was the biggest fundamental change for the Beta 4 release.  Instead of depending on the [Maquette](https://maquettejs.org/) virtual DOM library, we realised that our specialised requirements and needs were better solved in the medium and long term by integrating it directly.  So in Beta 4 we have replaced Maquette with `@dojo/widget-core/vdom`.
+
+This has allowed us to directly merge our abstraction of widgets and virtual DOM nodes with the virtual DOM nodes that get rendered to the DOM.  This increases the performance of the widgeting system, but also allowed us to make other performance and memory improvements to the widgeting system.
+
+This alone would not have had an effect on users of Dojo 2, but it allowed us to address a whole list of features we desired and refactoring of code that introduced some significant **breaking changes** in Beta 4.
+
+## Virtual DOM classes
+
+We felt that one of the biggest usability issues we had with Dojo 2 had to do with dealing with classes.  In Beta 4, we have changed this significantly, and while it is a **breaking change** we think the end result is _a lot_ more usable.
+
+Previously, `classes` were actually an object map of keys with a boolean value.  If you were using the theming system, you would use a function to specify your classes in a more functional way, which would generate a memoised map for you.  If you weren't you would have to do this yourself.  While it might be ok to author, it was really difficult to test.  It would look something like this:
+
+```typescript
+import { v } from '@dojo/widget-core/d';
+import WidgetBase from '@dojo/widget-core/WidgetBase';
+import ThemeableMixin, { theme } from '@dojo/widget-core/mixins/Themeable';
+import * as css from './styles/mywidget.m.css';
+
+const ThemeableBase = ThemeableMixin(WidgetBase);
+
+@theme(css)
+export default class MyWidget extends ThemeableBase {
+    protected render() {
+        return v('div', {
+            classes: this.classes(css.root).fixed(css.rootFixed)
+        }, [ 'Hello world!' ])
+    }
+}
+```
+
+With the changes in Beta 4, this gets written as:
+
+```typescript
+import { v } from '@dojo/widget-core/d';
+import WidgetBase from '@dojo/widget-core/WidgetBase';
+import ThemedMixin, { theme } from '@dojo/widget-core/mixins/Themed';
+import * as css from './styles/mywidget.m.css';
+
+const ThemedBase = ThemedMixin(WidgetBase);
+
+@theme(css)
+export default class MyWidget extends ThemedBase {
+    protected render() {
+        return v('div', {
+            classes: [ this.theme(css.root), css.rootFixed ]
+        }, [ 'Hello world!' ])
+    }
+}
+```
+
+The `classes` property is now just an array of strings.  Also we think that the mixed in method `theme()` is a lot more clear on what you are doing instead of `classes()`.  _Fixed_ classes are just simply added to the array, instead of being supplied to a chained function.
```

Comment:
_Fixed_ classes are just simply added to the array, instead of being supplied to a chained function. -> _Fixed_ classes are simply added to the array, instead of being supplied to a chained function.